### PR TITLE
fix the gradient stopping in EILoss

### LINKED ISSUE #262: Failing to stop the gradient in EILoss
The gradient stopping in `deepinv.loss.EILoss` is currently implemented using `torch.no_grad()`. Depending on how the transformations are implemented, this might not be enough and stopping the gradient might silently fail. More precisely, if the transformation function returns its own input, using `torch.no_grad()` has no effect and the resulting tensor is not marked as being independent of the learnable weights in the backpropagation algorithm. Instead, the transformed signal should be explicitely detached from the computational graph.

**Reproducing steps**

```python
import torch
from torch.nn import Module
from deepinv.transform import Shift
from deepinv.physics import Denoising
from deepinv.models import AutoEncoder, ArtifactRemoval

import torch
import torch.nn as nn
from deepinv.loss.loss import Loss


# the current EILoss with an additional assert statement
class EILoss(Loss):
    def __init__(
        self,
        transform,
        metric=torch.nn.MSELoss(),
        apply_noise=True,
        weight=1.0,
        no_grad=False,
    ):
        super(EILoss, self).__init__()
        self.name = "ei"
        self.metric = metric
        self.weight = weight
        self.T = transform
        self.noise = apply_noise
        self.no_grad = no_grad

    def forward(self, x_net, physics, model, **kwargs):
        if self.no_grad:
            with torch.no_grad():
                x2 = self.T(x_net)
            # make sure the gradient stopping has succeeded
            assert x2.is_leaf
        else:
            x2 = self.T(x_net)

        if self.noise:
            y = physics(x2)
        else:
            y = physics.A(x2)

        x3 = model(y, physics)

        loss_ei = self.weight * self.metric(x3, x2)
        return loss_ei

# unlike other transforms it may directly return its input which makes the gradient stopping fail
class Transform(Module):
    def __init__(self):
        super().__init__()
        self.T = Shift()

    def forward(self, x):
        if torch.rand(1) < .5:
            x = self.T(x)
        return x

loss = EILoss(Transform(), no_grad=True)

backbone = AutoEncoder(dim_input=1 * 32 * 32, dim_mid=128, dim_hid=32)
model = ArtifactRemoval(backbone)
physics = Denoising()
x = torch.randn((1, 1, 32, 32))
y = physics(x)
x_net = model(y, physics)

torch.manual_seed(0)
loss(y=y, x_net=x_net, physics=physics, model=model)
print("stopping the gradient has succeeded")

try:
    torch.manual_seed(1)
    loss(y=y, x_net=x_net, physics=physics, model=model)
except:
    print("stopping the gradient has failed")
```

**Expected behavior**

Both calls to the `loss` function should **not** fail with an exception.

**Actual behavior**

The first call to the `loss` function does not fail with an exception, but the second one does.

### PR DESCRIPTION
**Pull request by jscanvic**
Fixes #262 

### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Review by tachella — APPROVED** [review]
thanks for spotting this issue Jérémy - the fix looks good to me!
